scripts/generate_manifest.py

```diff
@@ -0,0 +1,128 @@
+from openpyxl import load_workbook
+import os
+
+from EPPs.common import SendMailEPP
+
+
+class GenerateManifest(SendMailEPP):
+    # populate the sample manifest with the sample date. Sample manifest template is determined by a step udf.
+    # The starting row and columns are determined by step UDFs. Uses SendMailEPP object for get_config function
+
+    _use_load_config = True  # should the config file be loaded?
+
+    # additional argument required to obtain the file location for newly create manifest in the LIMS step
+    def __init__(self, argv=None):
+        super().__init__(argv)
+        self.manifest = self.cmd_args.manifest
+
+    @staticmethod
+    def add_args(argparser):
+        argparser.add_argument(
+            '-m', '--manifest', type=str, required=True, help='Sample manifest generated by the LIMS'
+        )
+
+    def _run(self):
+
+        # obtain all of the inputs for the step
+        all_inputs = self.process.all_inputs(unique=True)
+        input_project_name=all_inputs[0].samples[0].project.name
+
+
+
+        # check all input containers have the same type and also create list of unique containers
+        container_types = set()
+        unique_container_names=set()
+
+
+        for artifact in all_inputs:
+            container_types.add(artifact.container.type)
+            unique_container_names.add(artifact.container.name)
+
+
+
+        if len(container_types) > 1:
+            raise ValueError('Only 1 container type is permitted. Multiple container types are present in the step')
+
+        # obtain step udfs
+        step_udfs = self.process.udf
+
+        # obtain the name of container type of the samples
+        if list(container_types)[0].name == '96 well plate':
+            con_type = '[Plates]'
+            template_file = self.get_config(config_heading_1='file_templates', config_heading_2='manifest',
+                                            config_heading_3='plate_template')
+        elif list(container_types)[0].name == 'rack 96 positions':
+            con_type = '[Tubes]'
+            template_file = self.get_config(config_heading_1='file_templates', config_heading_2='manifest', config_heading_3='tube_template')
+        elif list(container_types)[0].name == 'SGP rack 96 positions':
+            con_type = '[SGP]'
+            template_file = self.get_config(config_heading_1='file_templates', config_heading_2='manifest', config_heading_3='SGP_template')
+
+        # define counter to ensure each sample is written to a new well
+        row_counter = step_udfs[con_type + 'Starting Row']
+
+        # open the correct manifest template for the container type
+
+        wb = load_workbook(filename=template_file)
+        ws = wb.active
+
+        configurable_udfs = []
+
+        for step_udf_key in step_udfs.keys():
+            tag_position = step_udf_key.find('[Sample UDF]')
+
+            if tag_position > -1:
+                configurable_udfs.append(step_udf_key[tag_position + 12:])
```

Comment:
I would take that `configurable_udfs` list out of the function and make it a `property` of the class like 
```python
@property
def configurable_udfs(self):
    return [udf for udf in self.process.udf if udf.find('[Sample UDF]') > -1]
```
It seems that these udfs are mostly used with the tag anyway so there is no point in removing it.

Reply:
Updated